internal/pkg/vm/internal/vmcontext/actor_state_handle.go

```diff
@@ -77,35 +125,57 @@ func (h *actorStateHandle) UpdateRelease(obj interface{}) {
 		runtime.Abort("Storage commit error")
 	}
 
-	// make new head available to take
+	// update head
 	h.head = &newcid
+
+	// add validation
+	h.addCidValidation(obj, newcid)
+
+	return out, nil
+}
+
+func (h *actorStateHandle) addCidValidation(obj interface{}, expected cid.Cid) {
+	// checks that the cid of the obj has not changed
+	h.validations = append(h.validations, func() bool {
+		return h.cidOf(obj) == expected
+	})
 }
 
-// Release asserts that the state has not changed and releases the lock on it.
+// Validate validates that the state was mutated properly.
 //
-// No future calls to `Take` are allowed on this object.
-func (h *actorStateHandle) Release(obj interface{}) {
-	// Review: for @spec why are all this checks here?
-	// rt._checkRunning()
-	// rt._checkActorStateAcquired()
-	if h.head != nil {
-		runtime.Abort("Must call Take() before calling Release()")
+// This method is not part of the public API,
+// it is expected to be called by the runtime after each actor method.
+func (h *actorStateHandle) Validate() {
+	for _, vfn := range h.validations {
+		if !vfn() {
+			runtime.Abort("State mutated outside of Transaction() scope")
+		}
 	}
+}
 
+// Dragons: move this to the storage API
+func (h *actorStateHandle) cidOf(obj interface{}) cid.Cid {
+	// get Cid for object
+	// Dragons: this should not be storing
```

Comment:
This is easy enough to fix now. Factor the `if` block out of `storage.Put()` into an exported `storage.ToNode(interface{})` and call that.